tests/kernel/posix/pthread_rwlock/src/posix_rwlock.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <pthread.h>
+#include "ztest.h"
+
+#define N_THR 3
+#define STACKSZ 1024
+#define ONE_SECOND 1
+#define THREAD_PRIORITY 2
+
+K_THREAD_STACK_ARRAY_DEFINE(stacks, N_THR, STACKSZ);
+int exit_count;
+pthread_rwlock_t rwlock;
+
+static void *thread_top(void *p1)
+{
+	pthread_t pthread;
+	u32_t policy, ret = 0;
+	struct sched_param param;
+
+	pthread = (pthread_t) pthread_self();
+	pthread_getschedparam(pthread, &policy, &param);
+	printk("Thread %d scheduling policy = %d & priority %d started\n",
+	       (s32_t) p1, policy, param.priority);
+
+	ret = pthread_rwlock_tryrdlock(&rwlock);
+	if (ret) {
+		printk("Not able to get RD lock on trying, try again\n");
+		zassert_false(pthread_rwlock_rdlock(&rwlock),
+			      "Failed to acquire write lock");
+	}
+
+	printk("Thread %d got RD lock\n", (s32_t) p1);
+	sleep(1);
+	printk("Thread %d releasing RD lock\n", (s32_t) p1);
+	zassert_false(pthread_rwlock_unlock(&rwlock), "Failed to unlock");
+	printk("Thread %d acquiring WR lock\n", (s32_t) p1);
+	zassert_false(pthread_rwlock_wrlock(&rwlock),
+		      "Failed to acquire WR lock");
+	printk("Thread %d acquired WR lock\n", (s32_t) p1);
+	sleep(1);
+	printk("Thread %d releasing WR lock\n", (s32_t) p1);
+	zassert_false(pthread_rwlock_unlock(&rwlock), "Failed to unlock");
+	return NULL;
+}
+
+static void test_rw_lock(void)
+{
+	s32_t i, ret;
+	pthread_attr_t attr[N_THR];
+	struct sched_param schedparam;
+	pthread_t newthread[N_THR];
+	struct timespec time;
+
+	time.tv_sec = 1;
+	time.tv_nsec = 0;
+
+	zassert_false(pthread_rwlock_init(&rwlock, NULL),
+		      "Failed to create rwlock");
+
+	printk("\nmain acquire WR lock and 3 threads acquire RD lock\n");
+	ret = pthread_rwlock_trywrlock(&rwlock);
+
+	if (ret != 0) {
+		printk("Parent thread acquiring WR lock\n");
+		zassert_false(pthread_rwlock_timedwrlock(&rwlock, &time),
+			      "Failed to acquire write lock");
+	}
+
+	/* Creating N premptive threads in increasing order of priority */
+	for (i = 0; i < N_THR; i++) {
+		pthread_attr_destroy(&attr[i]);
+		pthread_attr_init(&attr[i]);
+
+		/* Setting scheduling priority */
+		schedparam.priority = i + 1;
+		pthread_attr_setschedparam(&attr[i], &schedparam);
+
+		/* Setting stack */
+		pthread_attr_setstack(&attr[i], &stacks[i][0], STACKSZ);
+
+		ret = pthread_create(&newthread[i], &attr[i], thread_top,
+				     (void *)i);
+		zassert_false(ret, "Low memory to thread new thread\n");
+
+	}
+
+	sleep(ONE_SECOND);
```

Comment:
why do we need these sleep statements? if you are trying to allow other threads to run, i would prefer you did this by setting thread priorities appropriately, or yielding the CPU, etc. having all these very long sleeps here just makes our CI runs take longer.

Reply:
Why sleep are provided :- 
1. Delay in main thread is provided to get child threads get a chance to run.
2. provided delay in between lock and unlock to make sure, even if thread is scheduled it keeps the lock.

But i got your point, we should not give too much delay as it delays QA execution. I have optimized the application from delay point. Not total delay in application is "<10 msec". please have a look.